plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgreInsertReplaceMethod.java

```diff
@@ -40,48 +42,82 @@ public class PostgreInsertReplaceMethod implements DBDInsertReplaceMethod {
-                            hasKey = true;
+                        getConstraintAttributesNames(constraintAttrNames, attributeReferences);
+                    }
+                }
+            }
+            if (constraintAttrNames.isEmpty()) {
+                // Let's search for an unique index
+                Collection<? extends DBSTableIndex> indexes = table.getIndexes(monitor);
+                if (!CommonUtils.isEmpty(indexes)) {
+                    Optional<? extends DBSTableIndex> optional = indexes.stream().filter(DBSTableIndex::isUnique).findFirst();
+                    if (optional.isPresent()) {
+                        DBSTableIndex tableIndex = optional.get();
+                        List<? extends DBSTableIndexColumn> references = tableIndex.getAttributeReferences(monitor);
+                        if (!CommonUtils.isEmpty(references)) {
+                            getConstraintAttributesNames(constraintAttrNames, references);
                         }
-                        StringBuilder updateExpression = new StringBuilder();
-                        updateExpression.append("(");
-                        addAttributesNamesList(table, attributes, false, updateExpression);
-                        updateExpression.append(") = (");
-                        addAttributesNamesList(table, attributes, true, updateExpression);
-                        updateExpression.append(")");
-                        query.append(" ").append(String.format(onConflictExpression, pkNames, updateExpression));
                     }
                 }
             }
+            if (!constraintAttrNames.isEmpty()) {
+                StringBuilder updateExpression = new StringBuilder();
+                updateExpression.append("(");
+                addAttributesNamesList(table, attributes, false, updateExpression);
+                updateExpression.append(") = (");
+                addAttributesNamesList(table, attributes, true, updateExpression);
+                updateExpression.append(")");
+                query.append(" ").append(String.format(onConflictExpression, constraintAttrNames, updateExpression));
+            } else {
+                log.debug("Can't find table constraints for the correct update on conflict operation.");
+                return null;
+            }
         } catch (DBException e) {
             log.debug("Can't read table constraints list", e);
         }
         return query.toString();
     }
 
-    private void addAttributesNamesList(DBSTable table, DBSAttributeBase[] attributes, boolean isExcluded, StringBuilder updateExpression) {
+    private void getConstraintAttributesNames(
+        @NotNull StringBuilder constraintAttrNames,
+        @NotNull List<? extends DBSEntityAttributeRef> attributeReferences
+    ) {
+        boolean hasKey = false;
+        for (DBSEntityAttributeRef column : attributeReferences) {
+            if (hasKey) constraintAttrNames.append(",");
```

Comment:
This should be placed after line 109; otherwise, if `attribute` is `null` for some reason, you can get multiple consecutive commas.